modules/replicator/src/testFixtures/java/org/apache/ignite/internal/raft/client/AbstractTopologyAwareGroupServiceTest.java

```diff
@@ -0,0 +1,522 @@
+
+                var commandsMarshaller = new ThreadLocalOptimizedMarshaller(cluster.serializationRegistry());
+
+                NodeOptions nodeOptions = new NodeOptions();
+                nodeOptions.setCommandsMarshaller(commandsMarshaller);
+
+                var raftServer = new JraftServerImpl(
+                        cluster,
+                        dataPath,
+                        nodeOptions,
+                        eventsClientListener
+                );
+                raftServer.start();
+
+                raftServer.startRaftNode(
+                        new RaftNodeId(GROUP_ID, localPeer),
+                        peersAndLearners,
+                        new TestRaftGroupListener(),
+                        RaftGroupOptions.defaults().commandsMarshaller(commandsMarshaller)
+                );
+
+                raftServers.put(addr, raftServer);
+
+                afterNodeStart(localPeer.consistentId(), cluster, dataPath, peersAndLearners, eventsClientListener);
+            }
+
+            if (addr.port() == clientPort) {
+                assertTrue(isServerAddress.test(addr));
+
+                raftClient = startTopologyAwareClient(cluster, clusterServices, isServerAddress, nodes, eventsClientListener, true);
+            }
+        }
+
+        return raftClient;
+    }
+
+    private TopologyAwareRaftGroupService startTopologyAwareClient(
+            ClusterService localClusterService,
+            Map<NetworkAddress, ClusterService> clusterServices,
+            Predicate<NetworkAddress> isServerAddress,
+            int nodes,
+            RaftGroupEventsClientListener eventsClientListener,
+            boolean notifyOnSubscription
+    ) {
+        if (eventsClientListener == null) {
+            eventsClientListener = new RaftGroupEventsClientListener();
+
+            var finalEventsClientListener = eventsClientListener;
+            localClusterService.messagingService().addMessageHandler(RaftMessageGroup.class, (msg, sender, correlationId) -> {
+                if (msg instanceof LeaderChangeNotification) {
+                    LeaderChangeNotification msg0 = (LeaderChangeNotification) msg;
+
+                    ClusterNode node = localClusterService.topologyService().getByConsistentId(sender);
+                    finalEventsClientListener.onLeaderElected(msg0.groupId(), node, msg0.term());
+                }
+            });
+        }
+
+        var commandsMarshaller = new ThreadLocalOptimizedMarshaller(localClusterService.serializationRegistry());
+
+        return TopologyAwareRaftGroupService.start(
+                GROUP_ID,
+                localClusterService,
+                FACTORY,
+                raftConfiguration,
+                peersAndLearners(clusterServices, isServerAddress, nodes),
+                true,
+                executor,
+                new LogicalTopologyServiceTestImpl(localClusterService),
```

Comment:
I cannot understand why we create `new LogicalTopologyServiceTestImpl(localClusterService)` here and in `org.apache.ignite.internal.placementdriver.ActiveActorTest#afterNodeStart`? Shouldn't be here only one `LogicalTopologyService`?

Reply:
fixed